docs/proposals/memory-management.md

```diff
@@ -0,0 +1,165 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/resource-qos.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Memory Management in Kubernetes
+
+**Author**: Vishnu Kannan ([@vishh](https://github.com/vishh))
+
+**Status**: Draft proposal
+
+#### Goals
+
+1. Document existing state of memory management.
+
+2. Explore opportunities for improving memory management in the future.
+
+#### Why is memory management important?
+
+Memory is a limited resource on machines. It is shared among multiple user containers and system processes. Unlike CPU, Memory is not compressible. Kubernetes being a cluster management system, aims to improve the utilization of memory on all nodes. Memory access latencies could vary a lot based on the architecture, state of the system, etc. When the available memory on a node becomes very low, the kernel can start killing processes. Kubernetes aims to provide a stable and predictable application performance by managing memory on the nodes, and providing efficient memory isolation for user containers. 
+
+
+#### How is memory managed as of now?
+
+###### Requirements
+
+* Cgroups are used to manage memory on the node.
+* Kernel Out of Memory killer is expected to be enabled. [`/proc/<pid>/oom_score_adj`](https://www.kernel.org/doc/Documentation/filesystems/proc.txt) is used to manage priority for access to memory. The kernel assigns an `oom_score` to individual processes based on the memory usage of the process and a few [other policies](https://lwn.net/Articles/317814/). `oom_score_adj` is taken into account while calculating the oom_score. Higher values for `oom_score_adj` will increase the likelihood for being considered for OOM kills.
+
+###### Node level memory management
+
+Kubelet, the node agent, performs some amount of memory management. Specifically, it can
+
+1. Expose the memory capacity and usage of the node.
+
+2. Apply memory limits to user containers and individual processes on the node.
+
+3. Set up cgroups to track the memory consumption of itself and other system processes like docker daemon, kube-proxy, sshd, etc. The cgroups being setup by kubelet are,
+  a. `/docker-daemon`: The docker daemon is under this cgroup. This cgroup is limited to 70% of the total available memory on the node.
+  b. `/system`: OS processes, excepting init, are under this cgroup.
```

Comment:
Systemd handling is broken as of now. @dchen1107 has been trying to come up with a `NodeSpec` and systemd configuration should be part of that. As of I assume this will translate to a services in system slice on a systemd machine, and kubelet does not understand the node setup.